dashboard/app/asset_storage.go

```diff
@@ -0,0 +1,347 @@
+// Copyright 2022 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package main
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/google/syzkaller/dashboard/dashapi"
+	"github.com/google/syzkaller/pkg/asset"
+	"golang.org/x/net/context"
+	"google.golang.org/appengine/v2"
+	db "google.golang.org/appengine/v2/datastore"
+	"google.golang.org/appengine/v2/log"
+)
+
+// TODO: decide if we really want to save job-related assets right now.
+//
+// The problem is that:
+// 1. We cannot first report the job status and then upload the assets -- an email will already
+//    be sent by that time.
+// 2. If we first upload assets and then construct a JobDoneReq response already with assets,
+//    we need to handle extra corner cases:
+//    - If JobDoneReq failed, assets must be deleted.
+//    - If we uploaded several files, all must be deleted.
+//    - What if syz-ci was killed somewhere in between? How do we get rid of those abadoned files?
+//
+// With the existing asset types it does not seem to be worth it to solve those problems.
+// Probably once we have things like code dumps it will start to make some sense.
+
+func saveBuildAsset(c context.Context, ns string, req *dashapi.AddBuildAssetReq) error {
+	asset := Asset{
+		Type:        asset.Type(req.AssetType),
+		DownloadURL: req.DownloadURL,
+		CreateDate:  timeNow(c),
+	}
+	tx := func(c context.Context) error {
+		build, err := loadBuild(c, ns, req.BuildID)
+		if err != nil {
+			return err
+		}
+		build.Assets = append(build.Assets, asset)
+		if _, err := db.Put(c, buildKey(c, ns, req.BuildID), build); err != nil {
+			return fmt.Errorf("failed to put build: %w", err)
+		}
+		log.Infof(c, "updated build: %#v", build)
+		return nil
+	}
+	if err := db.RunInTransaction(c, tx, &db.TransactionOptions{}); err != nil {
+		log.Errorf(c, "failed to update build: %v", err)
+		return err
+	}
+	return nil
+}
+
+func parallelize(errorStart string, funcs ...func() error) error {
+	errors := make(chan error)
+	for _, f := range funcs {
+		go func(callFunc func() error) {
+			errors <- callFunc()
+		}(f)
+	}
+	defer close(errors)
+	errorStr := ""
+	for i := 0; i < len(funcs); i++ {
+		err := <-errors
+		if err == nil {
+			continue
+		}
+		errorStr = fmt.Sprintf("%s%v\n", errorStr, err)
+	}
+	if errorStr != "" {
+		return fmt.Errorf("%s\n%s", errorStart, errorStr)
+	}
+	return nil
+}
+
+func queryNeededAssets(c context.Context) (*dashapi.NeededAssetsResp, error) {
+	// So far only build assets.
+	// TODO: once crash assets are implemented, parallelize the queries.
+	var builds []*Build
+	_, err := db.NewQuery("Build").
+		Filter("Assets.DownloadURL>", "").
+		GetAll(c, &builds)
```

Comment:
I thought about that, but it looks like it's just not worth it.

Why not per-namespace:
1) At the syz-ci level, we don't really know which namespace each syz-manager belongs. In the best case, we could group them by dashboard client and deprecate assets separately, but that looks somewhat excessive.
2) The current algorithm requires us to subtract the set of the existing assets and the needed ones. It means, we'd have to also store the assets separately, so that we would be able to only query GCS objects belonging to some specific namespace (and again -- we don't even know the namespace of each instance). Otherwise we'd just quickly end up with 0 remaining assets.
3) Once we implement crash-specific assets, it will be painful to query the assets from crashes, which are related to some specific namespace.

Storing them per-namespace looks a bit better -- at least we know the names of the managers, but (2) holds -- we'd need to also store them separately.

Another question is what performance benefits it can bring. If all syz-ci instances share the same GCS bucket, asset deprecation can be delegated to only one of them, so we won't really make a lot of unnecessary querying. And even if each instance queries them, it doesn't seem that we are going to have really a lot of assets.

We can probably prepend the manager name prefix before GCS object names (as "folders"), so that we could later query them per-manager, should it turn necessary. GCS API provides the ability to filter objects by path prefix: https://pkg.go.dev/cloud.google.com/go/storage#Query

WDYT?